tensorflow/c/c_api.cc

```diff
@@ -2399,9 +2405,24 @@ void TF_AddGradients(TF_Graph* g, TF_Output* y, int ny, TF_Output* x, int nx,
 
     const int first_new_node_id = g->graph.num_node_ids();
 
+    const char* child_scope_name = prefix;
+    if (child_scope_name != nullptr) {
+      // The operation should fail if the provided name prefix has already been
+      // used in this graph
+      for (const auto& pair: g->name_map) {
+        const string& name = pair.first;
+        if (name.compare(0, name.find_last_of('/'), prefix) == 0) {
+          status->status =
+              InvalidArgument("Duplicate node name in graph: '", prefix, "'");
+          return;
+        }
+      }
+    } else {
+      child_scope_name = "gradients";
+    }
     tensorflow::Scope scope =
         NewInternalScope(&g->graph, &status->status, &g->refiner)
-            .NewSubScope("gradients");
+            .NewSubScope(child_scope_name);
```

Comment:
This is a bit subtle - that we're relying on `NewSubScope` to adding the "_1", "_2" etc. when `child_scope_name == nullptr`, but *not* do so when a `prefix` is provided.

To ensure that this is the case, perhaps we should validate (around like 2445) that the added nodes meet the expectation that their name begins with `<prefix>/`?

Reply:
sounds good